src/core/inductive.ml

```diff
@@ -0,0 +1,257 @@
+        Print.pp_term (Bindlib.unbox res);
+    res
+  in
+
+  (* STEP 3: Create the conclusion of the inductive principle *)
+  let codom ts =
+    let x = Bindlib.new_var mkfree "x" in
+    let t = Bindlib.bind_var x (prf_of_p ts (_Vari x)) in
+    _Prod (fapp ind_sym ts) t
+  in
+  let conclusion = change_codom_ind_typ pos ind_sym codom in
+
+  (* STEP 4: Create the induction principle *)
+
+  let t =
+    List.fold_right (fun a b -> _Impl (case_of a) b) cons_list conclusion
+  in
+  let t = _Prod p_type (Bindlib.bind_var p t) in
+  if !log_enabled then
+    log_ind "The induction principle of the inductive type [%a] is %a"
+      Print.pp_term (Symb(ind_sym)) Print.pp_term (Bindlib.unbox t);
+  Bindlib.unbox t
+
+(** [type_to_pattern t] creates a pattern which matches with the type [t]. *)
+let rec type_to_pattern : term -> p_term = fun t ->
+  match t with
+  | Vari x    -> Pos.none (P_Patt (Some (Pos.none (Bindlib.name_of x)), None))
+  | Symb s       -> Pos.none (P_Iden (Pos.none ([], s.sym_name), true))
+  | Appl(t1, t2) -> Pos.none (P_Appl (type_to_pattern t1, type_to_pattern t2))
+  | Prod _       -> fatal None "Prod - Not yet implemented"
+  | Abst _       -> fatal None "Abst - Not yet implemented"
+  | LLet _       -> fatal None "LLet - Not yet implemented"
+  | Type         -> fatal None "TYPE - Not yet implemented"(*Pos.none P_Type*)
+  | Kind         -> fatal None "[Kind] not possible"
+  | Wild         -> fatal None "[Wild] not possible"
+  | TRef _       -> fatal None "[TRef] not possible"
+  | Meta _       -> fatal None "[Meta] not possible"
+  | Patt _       -> fatal None "[Patt] not possible"
+  | TEnv _       -> fatal None "[TEnv] not possible"
+
+(** Some constructors to create a p_term without position *)
+
+let _P_Iden : qident -> bool -> p_term = fun ident b ->
+  Pos.none (P_Iden(ident, b))
+
+let _P_Patt : strloc option -> p_term array -> p_term = fun s ts ->
+  Pos.none (P_Patt(s, Some ts))
+
+let _P_Appl : p_term -> p_term -> p_term = fun t1 t2 ->
+  Pos.none (P_Appl(t1, t2))
+
+(** [create_patt t] creates a pattern $[t] without position *)
+let create_patt : strloc -> p_term = fun t -> _P_Patt (Some t) [||]
+
+(*
+let create_arg : int -> p_term list = fun len ->
+  let rec aux : int -> p_term list -> p_term list = fun i acc ->
+    if i = 0 then
+      acc
+    else
+      let head =
+        Pos.none (P_Patt (Some (Pos.none ("a" ^ (string_of_int i))), None))
+      in
+      aux (i-1) (head::acc)
+  in
+  aux len []*)
+
+(** [rec_rules ind_sym rec_sym cons_list] returns the p_rules linking with an
+    induction principle [rec_sym] of the inductive type [ind_sym] (with its
+    constructors [cons_list]). *)
+let rec_rules : sym -> sym -> sym list -> p_rule list =
+  fun ind_sym rec_sym cons_list ->
+
+  (* STEP 0: Define some tools which will be useful *)
+  let app   : p_term -> p_term list -> p_term = List.fold_left _P_Appl in
+  let fapp  : p_term list -> p_term -> p_term = fun ts f -> app f ts in
```

Comment:
why f is after ts?? please change

Reply:
It's just to compute `ts` once:
```
let tmp = fapp (List.rev arg_list) in
              tmp t_ident, tmp p_patt
```

Reply:
This is not a very good reason! Instead, you should write `let args = List.rev arg_list in fapp t_ident args, fapp p_patt args`.

Reply:
Oh yes exact :')